compiler/src/iree/compiler/Dialect/Util/IR/UtilExternalModels.cpp

```diff
@@ -0,0 +1,150 @@
+  static void add(MLIRContext *ctx) {
+    add<OpTy1>(ctx);
+    add<OpTy2, More...>(ctx);
+  }
+};
+
+struct InsertSliceOpTiedOpInterface
+    : public TiedOpInterface::ExternalModel<InsertSliceOpTiedOpInterface,
+                                            tensor::InsertSliceOp> {
+  Value getTiedResult(Operation *op, unsigned resultIndex) const {
+    auto insertSliceOp = cast<tensor::InsertSliceOp>(op);
+    return IREE::Util::TiedOpInterface::findTiedBaseValue(
+        insertSliceOp.getDest());
+  }
+
+  ::llvm::Optional<unsigned> getTiedResultOperandIndex(
+      Operation *op, unsigned resultIndex) const {
+    return {1};  // dest
+  }
+
+  SmallVector<int64_t, 4> getTiedResultOperandIndices(Operation *op) const {
+    return {1};  // dest
+  }
+};
+
+template <typename OpTy>
+struct LinalgOpTiedOpInterface
+    : public TiedOpInterface::ExternalModel<LinalgOpTiedOpInterface<OpTy>,
+                                            OpTy> {
+  Value getTiedResult(Operation *op, unsigned resultIndex) const {
+    auto linalgOp = cast<OpTy>(op);
+    return IREE::Util::TiedOpInterface::findTiedBaseValue(
+        linalgOp.getOutputTensorOperands()[resultIndex]->get());
+  }
+
+  ::llvm::Optional<unsigned> getTiedResultOperandIndex(
+      Operation *op, unsigned resultIndex) const {
+    auto linalgOp = cast<OpTy>(op);
+    return {
+        linalgOp.getOutputTensorOperands()[resultIndex]->getOperandNumber()};
+  }
+
+  SmallVector<int64_t, 4> getTiedResultOperandIndices(Operation *op) const {
+    SmallVector<int64_t, 4> result;
+    for (unsigned i = 0; i < op->getNumResults(); ++i)
+      result.push_back(*getTiedResultOperandIndex(op, i));
+    return result;
+  }
+};
+
+/// Helper structure that iterates over all LinalgOps in `OpTys` and registers
+/// the `TiedOpInterface` with each of them.
+template <typename... Ops>
+struct LinalgOpTiedOpInterfaceHelper {
+  static void registerOpInterface(MLIRContext *ctx) {
+    (void)std::initializer_list<int>{
+        0, (Ops::template attachInterface<LinalgOpTiedOpInterface<Ops>>(*ctx),
+            0)...};
+  }
+};
+
+}  // namespace
+
+void registerUtilExternalModels(DialectRegistry &registry) {
+  // Must ensure that any dependent dialects are registered.
+  registry.insert<arith::ArithmeticDialect, linalg::LinalgDialect,
+                  tensor::TensorDialect>();
+
+  registry.addExtension(+[](MLIRContext *ctx,
+                            arith::ArithmeticDialect *dialect) {
+    GenericNumericCastExternalModel::add<
+        arith::BitcastOp, arith::ExtFOp, arith::ExtUIOp, arith::ExtSIOp,
+        arith::FPToSIOp, arith::FPToUIOp, arith::IndexCastOp, arith::TruncFOp,
+        arith::TruncIOp, arith::SIToFPOp, arith::UIToFPOp>(ctx);
+  });
+
+  registry.addExtension(+[](MLIRContext *ctx, tensor::TensorDialect *dialect) {
+    tensor::InsertSliceOp::attachInterface<InsertSliceOpTiedOpInterface>(*ctx);
+  });
+
+  registry.addExtension(+[](MLIRContext *ctx, linalg::LinalgDialect *dialect) {
```

Comment:
These are neat!
I'm a bit concerned about putting linalg stuff in Util/ as we generally try not to, but you've got this nicely separated into its own target and only pulled in by Tools/ so it's fine. Just something to think about if you find a place to put it that's closer to linalg-related things (linalgext/ etc) in the future.